### DISCUSSION TITLE
jsonable_encoder converts Decimal to float/int instead of str

### DISCUSSION THREAD

**Discussion by GianlucaFicarelli** [fastapi/fastapi#12050] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from decimal import Decimal
from fastapi.encoders import jsonable_encoder

pi = Decimal("3.14159265358979323846264338327950288419716939937510")
d = jsonable_encoder({"pi": pi, "one": Decimal("1")})
print(d)
# result: {'pi': 3.141592653589793, 'one': 1}

Decimal(d['pi']) == pi
# result: False
```

### Description

`jsonable_encoder()` converts Decimal to float/int instead of str, so it can cause loss of precision.

This behaviour is known because it's mentioned as a comment in:
https://github.com/fastapi/fastapi/blob/2d34086b70a1026aeb4974ffcf92041cd69e1ece/fastapi/encoders.py#L35-L55
but I didn't find a specific issue or discussion for this.

For reference, the discussion leading to serialize Decimal as a string in Pydantic v2 is at  https://github.com/pydantic/pydantic/issues/1511

### Operating System

macOS

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

3.12.1

**Comment by amacfie-tc** [comment]
Also, the [docs](https://fastapi.tiangolo.com/tutorial/extra-data-types/#other-data-types:~:text=%22format%22.-,Decimal,-%3A) say in requests and responses, `Decimal` is "handled the same as a `float`". But if you use a Pydantic v2 model in FastAPI it's serialized as a `str`.

**Comment by YuriiMotov** [comment, marked as answer]
You can define the custom encoder to change this:

```python
from decimal import Decimal
from fastapi.encoders import jsonable_encoder

pi = Decimal("3.14159265358979323846264338327950288419716939937510")
d = jsonable_encoder(
    {"pi": pi, "one": Decimal("1")},
    custom_encoder={Decimal: lambda v: str(v)},
)

print(d)
# result: {'pi': 3.14159265358979323846264338327950288419716939937510, 'one': 1}

print(Decimal(d['pi']) == pi)
# result: True

```

Or:
```python
from decimal import Decimal
from fastapi import FastAPI, Query
from pydantic import BaseModel, ConfigDict


app = FastAPI()


class ResModel(BaseModel):
    pi: Decimal
    one: Decimal

    model_config = ConfigDict(json_encoders={Decimal: lambda v: str(v)})


@app.get("/", response_model=ResModel)
async def index(d: Decimal = Query()):
    return {"pi": d, "one": Decimal("1")}

```

And without Pydantic model:

```python
from decimal import Decimal
from typing import Annotated, TypeAlias
from fastapi import FastAPI, Query
from pydantic import PlainSerializer


app = FastAPI()


MyDecimal: TypeAlias = Annotated[Decimal, PlainSerializer(str)]


@app.get("/", response_model=MyDecimal)
async def index(d: Decimal = Query()):
    return d

```
